### PR TITLE
feat: sessions and conversations view

### PR DESCRIPTION
**Pull request by peter-kismarczi**
## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [x] Unit tests for essential parts of your code, e.g. APIs exposed via SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person> for <code, bug, docs, etc>"
- [x] Linked issues 
    - #1822
    - #2010
    - #2009
    - #2007
    - #2006

# Summary
Monitoring UI for sessions and conversations. Sessions contain multiple conversations, conversations contain messages. Users can browse, filter, search, and interact with past and active sessions.

## Architecture
**Event-Sourced Design:**
- SessionsBroker maintains materialized view of sessions/queries/conversations
- Status, participants, and metadata computed server-side from Query CR events
- No pre-registration - participants appear after first query execution

**Data Sources:**
- **Sessions/Queries:** SessionsBroker (event-sourced from Query CRs)
- **Messages:** Memory Broker (OpenAI-format with tool calls, roles, sequence numbers)
- **Participants:** K8s API (agents, teams, tools)
- **Events/Logs:** Broker API (tool calls, query lifecycle events)

## Key Components

**Backend:**
- SessionsBroker: Cursor-based pagination, status computation, conversation aggregation
- Ark API: Proxy layer for sessions, conversations, messages, events

**Frontend:**
- Session list: Status filters, date range, search, sort (name/date/conversations)
- Session detail: Tabs for conversations, logs, with real-time updates
- Conversations: Message display, chat input, temporary conversation handling
- Cursor-based pagination throughout (sessions, logs)

## Test Coverage
- Service layer: broker-sessions, conversations APIs
- Component layer: SessionsTable, ConversationsTab, LogsTab, SessionDetailPage, MessageDisplay, NewSessionDialog, NewConversationDialog
- Page layer: Session detail page with async params, dialog interactions, message rendering
- Edge cases: pagination, filter combinations, state management, dialog state management, async component rendering
- All tests use mocked dependencies (unit test approach)

## Decisions

**Scope Decisions:**
- Token counts dropped - will implement when moved to broker
- Only single-participant sessions/conversations supported (agent, team, or tool)
- Conversation renaming deferred to future ticket
- Message counts use query count approximation (avoids additional Memory Broker calls)
- Tool call counts partially implemented via event scanning (may need refinement)
- Stuck with SessionsBroker + hybrid approach vs Query CRs only (pending data richness evaluation)

**Deferred Features:**
- Date range filter with dual-calendar picker (future enhancement)
- Attach items for messages (separate ticket)
- Reset chat functionality (separate ticket)
- SSE/real-time updates instead of polling (separate ticket)
- Participant and conversation activity status (currently only session-level status tracked)

### CODE REVIEW

**Comment by Nab-0** [comment]
1. Token Counting
Let's drop for now unless we move to borker

2. Multi-Participant Sessions
This shouldn't be the case, you can only select 1 participant, an agent, a team or a tool. 

3. Conversation Renaming
let's capture this in a ticket

4. Message Count Accuracy
keep it as the query count for now

5. Tool Call Count
yes, let's get this in please

6. Architecture Decision
Not sure - I think we should use whatever gives us richer data

**Comment by nataliatch** [comment]
SCREEN 1:

**Stats bar**

- [x] Formatting doesn't match the designs. 

designs image
implementation: 
image

**Filters above the table**

- [x] - Missing field labels above the dropdowns. The design has "Date range" above the first dropdown and "Status" above the second — currently they're unlabelled.
- [ ] - The first dropdown should be a Date range filter, not a generic "Choose option." 
pkismarczi: we don’t have proper dual-calendar date picker and also it might need some api/backend change as well. I put this at the end of the tasks list and would like to ask what if we create an issue for it?

**Table columns**

- [x] - The Convos header should have a sort arrow (↕) next to it, matching Name and Tokens.

**Session rows**

- [x] - Rows should support multiple participant tags in the Participants column (e.g. Data Collector + Bloomberg Terminal + Financial Analysis Team), not just one.
- [x] - Participant tags need distinct icons per type — agent, tool, and team each have their own icon in the design.
- [ ] - The count badge (e.g. "2", "10") next to the session name appears on multiple rows in the design but only on one row in the build (session-003).
pkismarczi: does it only show conversations with errors? if so, then it’s working [x]
- [x] - A "+N" overflow badge should appear at the end of a participants list when there are more than what fits inline (see "Financial Analysis Team +4" in the design).

designs: 
image

implementation: image

**Comment by nataliatch** [comment]
SESSION DETAILS: 

**Header / session metadata**

- [x] -  Session header card background colour is wrong — should be a subtle elevated surface (slightly lighter than page bg), implementation looks flat / same as page background
- [x] -  Header card padding is too tight — design has more generous internal padding (top/bottom and left/right), implementation feels cramped
- [x] -  Session ID (session-001 / sess_234567) font weight and size look off — design uses a heavier, larger treatment as the primary title
- [x] -  Vertical spacing between timestamp row, session ID row, and participant chips row is inconsistent with design
- [x] -  "Back to all sessions" chevron should be < not ←, inline with text, no boxed container
- [x] -  Status pill — design uses outlined pill ("Active"), implementation uses filled solid bg ("Idle"). Align to outlined treatment
- [x] -  Participant chips missing entirely — render a chip per participant (Data Collector, Alpha Analyst, Risk Assessor, Financial Analysis Team, Bloomberg Terminal), each with correct icon per type (agent vs team vs tool)
- [x] -  Chip padding, border radius, and border colour should match design spec (subtle border, rounded corners, consistent height)
- [x] -  Chips need status indicator dot (teal for active) and count badge for tools (e.g. Bloomberg Terminal 2)
- [ ] -  Delete/trash icon missing from right-hand side of header
pkismarczi: we will skip the deletion mechanism for this version
- [x] -  Vertical dividers between metadata items (Tokens | Conversations | Participants | errors) missing
- [x] -  Metadata row icons and labels need consistent gap spacing — implementation spacing looks uneven

**Tabs (History / Logs)**

- [x] -  Tab styling wrong — design uses underline indicator on active tab, implementation uses filled pill background. Switch to underline
- [x] -  Tabs should sit on a full-width bottom border
- [x] -  Padding around tab labels is too tight — needs more horizontal breathing room
- [x] -  Font weight on active vs inactive tab needs differentiation (active heavier)

**Conversations list (left panel)**

- [x] -  Left panel needs a right-edge divider separating it from the conversation detail area
- [x] -  Duration format inconsistent — design shows 12min / 2 min in a boxed pill at the bottom of each row; implementation shows 60m 0s inline. Align to design
- [x] -  Timestamps — design shows absolute times (14:12:13), implementation shows relative (2h, 1h).
- [x] -  Selected row needs left-edge accent border (teal for active selection, red for error state) — missing in implementation
- [x] -  Error indicator badge (red 2 pill) missing for conversations with errors
- [x] -  Participant icon should differ by type (agent vs human/team) — implementation uses same icon for all
- [x] -  Row padding (vertical and horizontal) is too tight in implementation — design has more generous spacing
- [x] -  Vertical gap between rows is inconsistent
- [x] -  Icon-to-text spacing within metadata row (icons + counts) needs tightening to match design
- [x] -  Row font sizes for participant name, timestamp, and metadata need to match design hierarchy

**Right panel (empty state)**

- [x] -  "Select a conversation" empty state is centred in the viewport — should be centred only within the right panel area, aligned to conversation list height
- [x] -  Empty state text colour and font size look off — should be a muted secondary text colour, not full white
- [x] @carolinastaudingermck can you please create an empty state for Peter?

**Right panel — populated state**

- [x] -  Panel header ("analyst-agent" + "Agent" badge) missing the bottom border that separates it from the message area in design
- [x] -  Agent" badge styling — design uses a subtle outlined/muted treatment, verify this matches
- [x] -  Header padding above and below the agent name row needs review
- [x]  Message input box styling differs:
- [x] - Design has the input flush within the panel with a subtle border on top
- [x] - Implementation shows the input as a separate boxed element with visible borders all around
- [x] - Send button in design is more compact; implementation's button feels oversized
- [ ] - Footer toolbar missing entirely below the input — design has attach icon, tool-call count badge, "Show tool calls" toggle on the left, and "Reset chat" action on the right
pkismarczi: attach file and reset chat functionalities are deferred
- [x] -  Input placeholder text styling (colour, weight) needs verification against design

**Global / theming**

- [x] -  Page background appears slightly off — verify against design token
- [x] -  Card / panel elevation not differentiated from page bg — design has subtle layering between page → card → nested elements
- [x] -  Border colours on cards and dividers too dark / not visible enough
- [x] -  Typography scale — title, body, metadata sizes need review against design spec
- [x] -  Icon sizes appear larger than design — verify against icon size tokens
- [x] -  Overall spacing scale (margins between header, tabs, content area) needs tightening

Implementation:
image
 Design: 
image

**Comment by nataliatch** [comment]
**Create new session:**

**Modal placement & layout**

- [x]  Modal width in implementation is wider than design — design is more compact / narrower
- [x]  Modal height scales to content in design — implementation looks taller with more empty space

**Header**

- [x]  Subtitle copy is wrong — implementation reads "Select a participant to start a session (agent, team, or tool)", design reads "Select one or more participants to start a session". Align to design
- [x]  Close (×) icon position — design has it inline with the title at the top right; implementation has it sitting in its own row above the title with extra spacing. Move it inline with the title
- [x]  Header padding above the title needs reducing to match design

**Participant list **

- [x]  Implementation shows "No participants found" empty state — design shows the populated list grouped by category with Agents, Teams, Tools
- [ ]  List items need checkboxes for multi-select (copy confirms "one or more")
pkismarczi: only single-participant sessions supported
- [x]  Each category needs a group header with name + count in parentheses (e.g. "Agents (7)", "Teams (2)")
- [x]  Each group has a subtle container / bordered background wrapping the checkbox rows together

Implementation: 
image
Design: 
image

**Comment by carolinastaudingermck** [comment]
Here is the requested empty state:

Screenshot 2026-04-28 at 14 03 14

File: services/ark-dashboard/ark-dashboard/components/sessions-conversations/conversation-sidebar.tsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  const date = new Date(timestamp);
+  return date.toLocaleTimeString('en-US', {
+    hour: '2-digit',
+    minute: '2-digit',
+    second: '2-digit',
+    hour12: false
+  });
+}
+
+function formatDurationPill(duration: string): string {
```
Comment (skazinka) [inline services/ark-dashboard/ark-dashboard/components/sessions-conversations/conversation-sidebar.tsx]: This doesn't seem to be used anywhere?

File: services/ark-dashboard/ark-dashboard/components/sessions-conversations/new-conversation-dialog.tsx [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+interface Props {
+  readonly open: boolean;
+  readonly onOpenChange: (open: boolean) => void;
+  readonly sessionParticipants: SessionParticipant[];
+  readonly selectedConversation: Conversation | null;
+  readonly onSelectParticipant: (participant: Participant) => void;
+}
+
+function getParticipantIcon(type: 'agent' | 'team' | 'tool') {
```
Comment (skazinka) [inline services/ark-dashboard/ark-dashboard/components/sessions-conversations/new-conversation-dialog.tsx]: There seem to be five almost but not exactly equal redefinitions of getParticipantIcon; seems to me it should be defined in one place and reused

File: services/ark-dashboard/ark-dashboard/components/sessions-conversations/session-tool-call.tsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  };
+  result?: string;
+}
+
+interface SessionToolCallProps {
+  toolCall: ToolCallData;
+  className?: string;
+}
+
+export function SessionToolCall({ toolCall, className }: Readonly<SessionToolCallProps>) {
```
Comment (skazinka) [inline services/ark-dashboard/ark-dashboard/components/sessions-conversations/session-tool-call.tsx]: This all seems really similar to the implementation in services/ark-dashboard/ark-dashboard/components/chat/tool-call.tsx, is there any deduplication we can do?

File: services/ark-dashboard/ark-dashboard/components/sessions-conversations/session-message.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+interface SessionMessageProps {
+  role: 'user' | 'assistant' | 'system';
+  content: string;
+  toolCalls?: ToolCallData[];
+  sender?: string;
+  timestamp?: string;
+  className?: string;
+}
+
+export function SessionMessage({
```
Comment (skazinka) [inline services/ark-dashboard/ark-dashboard/components/sessions-conversations/session-message.tsx:16]: Why not just reuse the existing chat message implementation in components/chat?

File: services/ark-dashboard/ark-dashboard/components/sessions-conversations/conversation-sidebar.tsx (hunk 1 above)
Comment (peter-kismarczi) [reply services/ark-dashboard/ark-dashboard/components/sessions-conversations/conversation-sidebar.tsx]: good catch, I'll remove the function

File: services/ark-dashboard/ark-dashboard/components/sessions-conversations/new-conversation-dialog.tsx (hunk 2 above)
Comment (peter-kismarczi) [reply services/ark-dashboard/ark-dashboard/components/sessions-conversations/new-conversation-dialog.tsx]: valid, I'll create an util function for this

File: services/ark-dashboard/ark-dashboard/components/sessions-conversations/session-tool-call.tsx (hunk 3 above)
Comment (peter-kismarczi) [reply services/ark-dashboard/ark-dashboard/components/sessions-conversations/session-tool-call.tsx]: yes, I'll modify the existing ToolCall function to handle the newly created one as well

File: services/ark-dashboard/ark-dashboard/components/sessions-conversations/session-message.tsx (hunk 4 above)
Comment (peter-kismarczi) [reply services/ark-dashboard/ark-dashboard/components/sessions-conversations/session-message.tsx:16]: the only truly common part was the ToolCall function, the other components are quite different (e.g. message display, scroll detection), so I would keep them separate for now.
if we decide to unify these chat messages in the new design, we can refactor this later

Comment (skazinka) [reply services/ark-dashboard/ark-dashboard/components/sessions-conversations/session-message.tsx:16]: I guess that's fair, I'm just curious why they're so different

**Comment by peter-kismarczi** [comment]
@ark review this PR

**Comment by hbachala** [comment]
## UI Test Results — Sessions & Conversations View

**26/26 tests passed** — all sessions UI features verified on local cluster.

### Test Coverage

| Area | Tests | Status |
|------|-------|--------|
| Sessions list page loads | 1 | ✅ |
| New Session button | 1 | ✅ |
| Search input | 1 | ✅ |
| Status filter (Active, Error, All) | 4 | ✅ |
| Date range filter (24h, 7d) | 2 | ✅ |
| Sort controls (Name, Convos) | 3 | ✅ |
| Sort direction toggle | 1 | ✅ |
| Search + clear | 2 | ✅ |
| New Session dialog (open, search, tabs, agents) | 4 | ✅ |
| Session detail page navigation | 2 | ✅ |
| Session detail tabs (History, Logs) | 3 | ✅ |
| New conversation button | 1 | ✅ |
| Back button navigation | 1 | ✅ |

### Observations

- Sessions list loads at `/session-history` with 41 sessions in the broker
- Status filters: All / Active / Idle / Error work correctly
- Date filters: Last 24h / Last 7 days applied correctly
- Sort by Name and Convos with direction toggle work
- New Session dialog shows agent/team/tool tabs with search
- Session detail renders History and Logs tabs
- Back button navigates to `/session-history`

### Minor Findings

- **No "Date" sort button in the UI** — sort field `date` is the default but has no header button (only Name and Convos have sort buttons). Consider adding a Date sort button for completeness.
- **Tab label is "History" not "Conversations"** — the conversations tab in session detail is labeled "History". Consistent with the data shown but may be worth noting if the PR description says "Conversations tab".
- **New conversation is an icon-only button (Plus icon)** — no text label. Could benefit from a tooltip or `aria-label` for accessibility.

Tested on cluster with 41 existing sessions including active, idle, and error states.

**Comment by peter-kismarczi** [comment]
@ark review this PR again

**Comment by peter-kismarczi** [comment]
@ark review the PR again

**Review by Nab-0 — APPROVED** [review]
